ffi/src/lib.rs

```diff
@@ -40,3 +54,741 @@ impl Iterator for EngineIterator {
         }
     }
 }
+
+/// A non-owned slice of a UTF8 string, intended for arg-passing between kernel and engine. The
+/// slice is only valid until the function it was passed into returns, and should not be copied.
+///
+/// # Safety
+///
+/// Intentionally not Copy, Clone, Send, nor Sync.
+///
+/// Whoever instantiates the struct must ensure it does not outlive the data it points to. The
+/// compiler cannot help us here, because raw pointers don't have lifetimes. To reduce the risk of
+/// accidental misuse, it is recommended to only instantiate this struct as a function arg, by
+/// converting a `&str` value `Into<KernelStringSlice>`, so the borrowed reference protects the
+/// function call (callee must not retain any references to the slice after the call returns):
+///
+/// ```
+/// fn wants_slice(slice: KernelStringSlice) { ... }
+/// let msg = String::from(...);
+/// wants_slice(msg.as_ref().into());
+/// ```
+#[repr(C)]
+pub struct KernelStringSlice {
+    ptr: *const c_char,
+    len: usize,
+}
+
+// Intentionally not From, in order to reduce risk of accidental misuse. The main use is for callers
+// to pass e.g. `my_str.as_str().into()` as a function arg.
+#[allow(clippy::from_over_into)]
+impl Into<KernelStringSlice> for &str {
+    fn into(self) -> KernelStringSlice {
+        KernelStringSlice {
+            ptr: self.as_ptr().cast(),
+            len: self.len(),
+        }
+    }
+}
+
+trait TryFromStringSlice: Sized {
+    unsafe fn try_from_slice(slice: KernelStringSlice) -> DeltaResult<Self>;
+}
+
+impl TryFromStringSlice for String {
+    /// Converts a slice back to a string
+    ///
+    /// # Safety
+    ///
+    /// The slice must be a valid (non-null) pointer, and must point to the indicated number of
+    /// valid utf8 bytes.
+    unsafe fn try_from_slice(slice: KernelStringSlice) -> DeltaResult<String> {
+        let slice = unsafe { std::slice::from_raw_parts(slice.ptr.cast(), slice.len) };
+        match std::str::from_utf8(slice) {
+            Ok(s) => Ok(s.to_string()),
+            Err(e) => Err(Error::generic_err(e)),
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub enum KernelError {
```

Comment:
this (and the `From` below) feel like something that should be derived via a proc macro. it's a bit tricky because we don't want it to live in the same module that we define our `deltakernel::Error` variant. I suspect there's something we could do though. Probably fine to have it be future work.

Reply:
https://github.com/delta-incubator/delta-kernel-rs/issues/161